strawberry/annotation.py

```diff
@@ -0,0 +1,261 @@
+        if not isinstance(other, StrawberryAnnotation):
+            return NotImplemented
+
+        return self.resolve() == other.resolve()
+
+    def resolve(self) -> Union[StrawberryType, type]:
+        annotation: object
+        if isinstance(self.annotation, str):
+            annotation = ForwardRef(self.annotation)
+        else:
+            annotation = self.annotation
+
+        evaled_type = _eval_type(annotation, self.namespace, None)
+        if evaled_type is None:
+            raise ValueError("Annotation cannot be plain None type")
+        if self._is_async_generator(evaled_type):
+            evaled_type = self._strip_async_generator(evaled_type)
+        if self._is_lazy_type(evaled_type):
+            return evaled_type
+
+        if self._is_generic(evaled_type):
+            if any(is_type_var(type_) for type_ in evaled_type.__args__):
+                return evaled_type
+            return self.create_concrete_type(evaled_type)
+
+        # Simply return objects that are already StrawberryTypes
+        if self._is_strawberry_type(evaled_type):
+            return evaled_type
+
+        # Everything remaining should be a raw annotation that needs to be turned into
+        # a StrawberryType
+        if self._is_enum(evaled_type):
+            return self.create_enum(evaled_type)
+        if self._is_list(evaled_type):
+            return self.create_list(evaled_type)
+        elif self._is_optional(evaled_type):
+            return self.create_optional(evaled_type)
+        elif self._is_scalar(evaled_type):
+            return evaled_type
+        elif self._is_union(evaled_type):
+            return self.create_union(evaled_type)
+        elif is_type_var(evaled_type):
+            return self.create_type_var(evaled_type)
+
+        # TODO: Raise exception now, or later?
+        # ... raise NotImplementedError(f"Unknown type {evaled_type}")
+        return evaled_type
+
+    def create_concrete_type(self, evaled_type: type) -> type:
+        if _is_object_type(evaled_type):
+            type_definition: TypeDefinition
+            type_definition = evaled_type._type_definition  # type: ignore
+            return type_definition.resolve_generic(evaled_type)
+
+        raise ValueError(f"Not supported {evaled_type}")
+
+    def create_enum(self, evaled_type: Any) -> EnumDefinition:
+        return evaled_type._enum_definition
+
+    def create_list(self, evaled_type: Any) -> StrawberryList:
+        of_type = StrawberryAnnotation(
+            annotation=evaled_type.__args__[0],
+            namespace=self.namespace,
+        ).resolve()
+
+        return StrawberryList(of_type)
+
+    def create_optional(self, evaled_type: Any) -> StrawberryOptional:
+        types = evaled_type.__args__
+        non_optional_types = tuple(
+            filter(lambda x: x is not type(None), types)  # noqa: E721
+        )
+
+        # Note that this returns _not_ a Union if len(non_optional_types) == 1
```

Comment:
this comment is strangely worded :D

Reply:
It's definitely a bit strange, but I couldn't think of a better phrasing that didn't also introduce some ambiguity

Reply:
I'd do something like this:
```
# Note that passing a single type to `Union` is equivalent to not using `Union` at all
# so here we don't do any checks for how many types have been passed as we can 
# safely use `Union` for both optional types (eg `Optional[str]`) and optional
# unions (eg `Optional[Union[TypeA, TypeB]]`)
```